sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java

```diff
@@ -57,7 +58,10 @@
    * Maps the external jar location to the functions the jar defines. Static so it can persist
    * across multiple SQL transforms.
    */
-  private static final Map<String, FunctionDefinitions> cache = new HashMap<>();
+  private static final Map<String, FunctionDefinitions> functionCache = new HashMap<>();
+
+  /** Maps potentially remote jar paths to their local file copies. */
+  private static final Map<String, File> jarCache = new HashMap<>();
```

Comment:
To confirm: for the created random files, as I recall they will live while the JVM is alive?

Just to confirm that there will not be the case the there is a entry in this jarCache but the jar file does not exist.

Reply:
I can only think of a couple ways the cache could contain handles to nonexistent files:

- If the file was deleted somehow.
- If the cache was serialized and shipped to workers. But we don't include a JavaUdfLoader instance as part of CalcFn, instead we construct a new JavaUdfLoader instance in `CalcFn::setup`. Besides, the cache is a static variable, and static variables are never serialized.

Reply:
I see. 

I see there are IOException thrown when loading jars. As long as there are meaningful exception are caught, at least we will have exceptions in log.